BAU: Simplify statemachine definition

Wynndow opened the pull request:
<!-- Provide a general summary of your changes in the Title above -->
<!-- Include the Jira ticket number in square brackets as prefix, eg `[P4-XXXX] PR Title` -->

## Proposed changes

### What changed

Simplify statemachine definition

### Why did it change
This commit is an attempt to make the statemachine definition less verbose and easier to understand. It does a few things.

This commit is an attempt to make the statemachine definition less
verbose and easier to understand. It does a few things.

* It removes the "name" property from states and events. These are now
  interpreted from the key used to define the body of them in the yaml.
  This used to set the name value on the objects as we still need it.
* It moves the "response" definition (effectively how to produce the
  desired state), from the event to the state itself. This makes things
  easier to reason about and prevents a bunch of duplication.
* It removes some duplicated states in favour of a shared one.
* Removes the StateMachineResult class as it's no longer needed. We can
  just use the state itself and the response it contains.
* Removes the criEvent type. It's not really needed now as States can
  have a response of type criResponse. This does push some logic into
  the basic event, but it means the definition is simpler.

This has been deployed and it works, but only with the new statemachine
files. If this change was to be applied, it would need to go in after
we've switched.

This has been deployed and it works, but only with the new statemachine files. If this change was to be applied, it would need to go in after we've switched.

Wynndow committed:
BAU: Simplify statemachine definition

This commit is an attempt to make the statemachine definition less
verbose and easier to understand. It does a few things.

* It removes the "name" property from states and events. These are now
  interpreted from the key used to define the body of them in the yaml.
  This used to set the name value on the objects as we still need it.
* It moves the "response" definition (effectively how to produce the
  desired state), from the event to the state itself. This makes things
  easier to reason about and prevents a bunch of duplication.
* It removes some duplicated states in favour of a shared one.
* Removes the StateMachineResult class as it's no longer needed. We can
  just use the state itself and the response it contains.
* Removes the criEvent type. It's not really needed now as States can
  have a response of type criResponse. This does push some logic into
  the basic event, but it means the definition is simpler.

This has been deployed and it works, but only with the new statemachine
files. If this change was to be applied, it would need to go in after
we've switched.

Wynndow committed:
BAU: Remove attempt-recovery event

This adds a check on the state to direct any "attempt-recovery" event
back to itself. This means we no longer need to define that event
repetitively in the statemachine.

Wynndow committed:
BAU: Use DEDUCTION when deserializing events

Using DEDUCTION instructs the objectmapper to use the fields to
determine the sub type to deserialize to, rather than having to specify
the "type" field as we currently do. This means we can strip out the
type from all the events in the statemachine file.

lwebb-iig reviewed (approved):
I generally think a lot of the changes here are well worth merging as it significantly reduces redundant & duplicated information in the statemachine files. If we consider the original goal of the refactor epic and the long term goals of having the statemachine files more readable and representative of the UX/User journey, then these changes certainly take a large step towards that goal.

lwebb-iig reviewed (changes requested):
Needs a spotless check

Wynndow committed:
BAU: Keep statemachine file up to date

This comes from a rebase on main which had new states added that needed
changing to the new format.
